### DISCUSSION TITLE
py03 how to use scipy csc matrix

### DISCUSSION THREAD

**Discussion by Liripo** [PyO3/pyo3#4295] [Questions]
I would like to know how to use csc matrix in pyo3, for example, how to convert it into sprs CsMat?

**Comment by flying-sheep** [comment]
nothing yet I think: https://github.com/sparsemat/sprs/issues/289

**Comment by ilgrad** [comment]
You can bridge the CSC storage arrays explicitly. SciPy and `sprs` both represent a CSC matrix using its shape, column pointers, row indices, and values.

First canonicalize a copy on the Python side:

```python
A = A.tocsc(copy=True)
A.sum_duplicates()
A.sort_indices()
```

Then pass `A.shape`, `A.indptr`, `A.indices`, and `A.data` to your PyO3 function. For an owned real-valued matrix, convert them into `(usize, usize)`, `Vec<usize>`, `Vec<usize>`, and `Vec<f64>`, respectively, and construct:

```rust
CsMat::<f64>::try_new_csc(shape, indptr, indices, data)
```

Use the fallible constructor and translate its error into a Python exception. SciPy index arrays may be `int32` or `int64`, so handle both with checked conversion to the Rust index type. Likewise, explicitly support or reject the input value dtype; the example above assumes real `f64` values.

This creates an owned copy. A borrowed `CsMatView` needs additional lifetime and mutation guarantees for the Python buffers.

References: [SciPy CSC storage](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html), [sprs constructors](https://docs.rs/sprs/latest/sprs/struct.CsMatBase.html#method.try_new_csc).
